File: 0062_unique_paths/unique_paths.js
/**
 * Time:    O(m * n)
 * Space:   O(m * n)
 * @param   {number} m
 * @param   {number} n
 * @returns {number}
 */
const uniquePaths = (m, n) => {
  const dp = []
  let tmp = m
  while (tmp--) dp.push(Array(n).fill(1))
  for (let i = 1; i < m; i++)
    for (let j = 1; j < n; j++) dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
  return dp[m - 1][n - 1]
}

File: 0062_unique_paths/unique_paths.go
package main

/**
 * Time: O(m * n)
 * Space: O(m * n)
 */
func uniquePaths(m int, n int) int {
	dp := make([][]int, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]int, m)
	}
	for r := 0; r < n; r++ {
		for c := 0; c < m; c++ {
			if r == 0 || c == 0 {
				dp[r][c] = 1
			} else {
				dp[r][c] = dp[r-1][c] + dp[r][c-1]
			}
		}
	}
	return dp[n-1][m-1]
}

func main() {

}
